partiql-parser/src/scanner.rs

```diff
@@ -0,0 +1,270 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! Provides a simple API to scan PartiQL syntax.
+//!
+//! The [`Scanner`] trait provides tools the capability to recognize PartiQL lexemes such as
+//! keywords, identifiers, etc.  This API is not a full parser, but a lexer in
+//! traditional parser terminology.  This API can be useful for tooling or IDEs that wish to
+//! do things like syntax highlighting, where full parsing is not required.
+
+use crate::peg::{PairExt, PairsExt, PartiQLParser, Rule};
+use crate::prelude::*;
+use crate::result::syntax_error;
+use pest::iterators::Pair;
+use pest::Parser;
+
+/// The parsed content associated with a [`Token`] that has been scanned.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Content<'val> {
+    /// A PartiQL keyword.  Contains the slice for
+    Keyword(&'val str),
+    // TODO things like literals, punctuation, etc.
+}
+
+/// Internal type to keep track of remaining input and relative line/column information.
+///
+/// This is used to leverage the PEG to do continuation parsing and calculating the line/offset
+/// information correctly.  Line/offset information does not correspond to UTF-8 code unit (octet)
+/// positions so this has to be tracked separately.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+struct Remainder<'val> {
+    /// The remainder of text to scan.
+    input: &'val str,
+    /// The offset in the input translate locations from
+    offset: LineAndColumn,
+}
+
+impl<'val> Remainder<'val> {
+    /// Produces a new [`Remainder`] by slicing the input by the given amount and providing
+    /// a new line/column baseline.
+    ///
+    /// The offset given is the logical one as if the input was the start of the sequence.
+    /// This method will calculate the new [`Remainder`] offset based on that.
+    fn consume(&self, amount: usize, offset: LineAndColumn) -> Self {
+        Self {
+            input: &self.input[amount..],
+            offset: self.offset.move_relative(offset),
+        }
+    }
+}
+
+/// A lexeme of PartiQL derived from a slice of input text.
+///
+/// A token can be thought of as a sort of continuation, it knows where it is in the input list
+/// and allows scanning to resume from it.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Token<'val> {
+    /// The underlying value of the token.
+    content: Content<'val>,
+    /// Start location for this token.
+    start_location: LineAndColumn,
+    /// End location for this token.
+    end_location: LineAndColumn,
```

Comment:
Should these be in a struct by themselves?

```
struct Span {
    /// Start location for this token.
    start_location: LineAndColumn,
    /// End location for this token.
    end_location: LineAndColumn,
}
```
??

Reply:
They could be--though the question is to what end in the API? Do you want to see `Token::span() -> Span`

Reply:
Added #14  to track this.

Reply:
`Token::span() -> Span` is the idea.